src/_pytest/fixtures.py

```diff
@@ -1679,3 +1679,17 @@ def _matchfactories(
         for fixturedef in fixturedefs:
             if fixturedef.baseid in parentnodeids:
                 yield fixturedef
+
+
+def pytest_runtest_teardown(item: nodes.Item, nextitem: nodes.Item) -> None:
+    if (
+        type(item).__name__ != "TestCaseFunction"
+        and hasattr(item, "_request")  # For YamlItem in testing/
+        and nextitem
+    ):
+        for fixture_name, fixture_def in item._request._fixture_defs.items():
+            if (
+                fixture_def._scope in HIGH_SCOPES
+                and fixture_name not in nextitem._request.fixturenames  # type: ignore[attr-defined]
```

Comment:
That condition is incorrect as it ignores cases where 2 tests using a fixture are separated by a test not using it

Additionally it ignores the case where a interspersed test uses a derivation of the fixture, thus using the fixture indirectly

Reply:
to elaborate
```python
import pytest

HACK = []


@pytest.fixture(scope='module')
def a()
    HACK.append(a)
    return "a"

@pytest.fixture(scope='module')
def b():
    HACK.append(b)
    return "b"

def test_a(a):
    assert a == 'a'

def test_b(b):
    assert b == 'b'

def test_c(a):
    assert True

def test_d(b):
   pass
   
def test_setups():
   assert HACK == [a, b]
```
should hold true, with just next-item it cant

Reply:
This would hold true in the new solution.